opentasks-provider/src/main/java/org/dmfs/provider/tasks/processors/instances/TaskValueDelegate.java

```diff
@@ -0,0 +1,167 @@
+ *
+ * @author Marten Gajda
+ */
+public final class TaskValueDelegate implements EntityProcessor<InstanceAdapter>
+{
+
+    private final EntityProcessor<TaskAdapter> mDelegate;
+
+
+    public TaskValueDelegate(EntityProcessor<TaskAdapter> delegate)
+    {
+        mDelegate = delegate;
+    }
+
+
+    @Override
+    public InstanceAdapter insert(SQLiteDatabase db, InstanceAdapter entityAdapter, boolean isSyncAdapter)
+    {
+        // for now inserts are just forwarded, but we have to make sure we return an InstanceAdapter which contains the current instance row _ID.
+        // TODO: in case of a recurring master we should add an RDATE to it
+        // TODO: handle when inserting an override that already exists, either update the existing one or fail
+        TaskAdapter taskResult = mDelegate.insert(db, entityAdapter.taskAdapter(), false);
+        Cursor c = db.query(TaskDatabaseHelper.Tables.INSTANCES, new String[] { TaskContract.Instances._ID },
+                TaskContract.Instances.TASK_ID + "=" + taskResult.id(), null, null, null, null);
+        try
+        {
+            c.moveToFirst();
+            return new ContentValuesInstanceAdapter(c.getLong(0), new ContentValues());
+        }
+        finally
+        {
+            c.close();
+        }
+    }
+
+
+    @Override
+    public InstanceAdapter update(SQLiteDatabase db, InstanceAdapter entityAdapter, boolean isSyncAdapter)
+    {
+        // if this is the master of a recurring task, we create a new instance for this override, otherwise we just delegate
+        TaskAdapter taskAdapter = entityAdapter.taskAdapter();
+        if (taskAdapter.isRecurring())
+        {
+            // clone the task to create an unsynced override
+            InstanceAdapter newInstanceAdapter = entityAdapter.duplicate();
+            TaskAdapter task = newInstanceAdapter.taskAdapter();
+            task.set(TaskAdapter.ORIGINAL_INSTANCE_ID, entityAdapter.valueOf(InstanceAdapter.TASK_ID));
+            task.set(TaskAdapter.ORIGINAL_INSTANCE_TIME, entityAdapter.valueOf(InstanceAdapter.INSTANCE_ORIGINAL_TIME));
+            task.unset(TaskAdapter.SYNC1);
+            task.unset(TaskAdapter.SYNC2);
+            task.unset(TaskAdapter.SYNC3);
+            task.unset(TaskAdapter.SYNC4);
+            task.unset(TaskAdapter.SYNC5);
+            task.unset(TaskAdapter.SYNC6);
+            task.unset(TaskAdapter.SYNC7);
+            task.unset(TaskAdapter.SYNC8);
+            task.unset(TaskAdapter.SYNC_ID);
+            task.unset(TaskAdapter.SYNC_VERSION);
+            // TODO: if this is the first instance (and maybe no other overrides exist), don't create an override but split the series into two tasks
+            mDelegate.insert(db, task, false);
+        }
+        else
+        {
+            // this is a non-recurring task or it's already an override, just delegate the update
+            mDelegate.update(db, taskAdapter, false);
```

Comment:
What about a direct update to a recurring instance? Wouldn't that lead to here and by simply changing it, braking the data consistency of the recurrence set?
Shouldn't that be detected here if it is a recurring instance?
Or are those going to have `RRULE` or `RDATE` copied from the master, so those are handled by the `if` branch here?

Reply:
Not sure what exactly you mean. A recurring instance (which is not an override) will always point to the master task. Which means they are handled by the `if` branch.

Reply:
Ok, I see now. So I assume 'pointing to the master task' means that the row of them in `Instance_View` are joined on master task's `Task` table row.

So in `Tasks` table we will only have the:
- non-recurring tasks
- recurring masters
- recurring overrides
but no rows for the enumerated recurring instances, right?

Reply:
Exactly. The enumerated recurrence instances go to the instances table, which essentially represents the instances as observed by the task UI. Since all non-overridden instances share the same data (except for start and due) it wouldn't make sense to copy the values over and over again.

Reply:
Clear, thanks.